networkoperator/v1/types_egressrouter.go

```diff
@@ -0,0 +1,274 @@
+}
+
+// ProtocolType defines the protocol types that are supported
+type ProtocolType string
+
+const (
+	// ProtocolTypeTCP refers to the TCP protocol
+	ProtocolTypeTCP ProtocolType = "TCP"
+
+	// ProtocolTypeUDP refers to the UDP protocol
+	ProtocolTypeUDP ProtocolType = "UDP"
+
+	// ProtocolTypeSCTP refers to the SCTP protocol
+	ProtocolTypeSCTP ProtocolType = "SCTP"
+)
+
+// EgressRouterInterface contains the configuration of interface to create/use.
+type EgressRouterInterface struct {
+	// Arguments specific to the interfaceType macvlan
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:default:={mode: Bridge}
+	Macvlan MacvlanConfig `json:"macvlan" protobuf:"bytes,1,opt,name=macvlan"`
+}
+
+// MacvlanMode defines the different types of modes that are supported for the macvlan interface.
+// source: https://man7.org/linux/man-pages/man8/ip-link.8.html
+type MacvlanMode string
+
+const (
+	// MacvlanModeBridge connects all endpoints directly to each other, communication is not redirected through the physical interface's peer.
+	MacvlanModeBridge MacvlanMode = "Bridge"
+
+	// MacvlanModePrivate does not allow communication between macvlan instances on the same physical interface,
+	// even if the external switch supports hairpin mode.
+	MacvlanModePrivate MacvlanMode = "Private"
+
+	// MacvlanModeVEPA is the Virtual Ethernet Port Aggregator mode. Data from one macvlan instance to the other on the
+	// same physical interface is transmitted over the physical interface. Either the attached switch needs
+	// to support hairpin mode, or there must be a TCP/IP router forwarding the packets in order to allow
+	// communication. This is the default mode.
+	MacvlanModeVEPA MacvlanMode = "VEPA"
+
+	// MacvlanModePassthru mode gives more power to a single endpoint, usually in macvtap mode.
+	// It is not allowed for more than one endpoint on the same physical interface. All traffic will be forwarded
+	// to this endpoint, allowing virtio guests to change MAC address or set promiscuous mode in order to bridge the
+	// interface or create vlan interfaces on top of it.
+	MacvlanModePassthru MacvlanMode = "Passthru"
+)
+
+// MacvlanConfig consists of arguments specific to the macvlan EgressRouterInterfaceType
+type MacvlanConfig struct {
+	// Mode depicts the mode that is used for the macvlan interface; one of Bridge|Private|VEPA|Passthru. The default mode is "Bridge".
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:Enum="Bridge";"Private";"VEPA";"Passthru"
+	// +kubebuilder:default:="Bridge"
+	Mode MacvlanMode `json:"mode" protobuf:"bytes,1,opt,name=mode,casttype=MacvlanMode"`
+
+	// Name of the master interface. Need not be specified if it can be inferred from the IP address.
+	// +kubebuilder:validation:MinLength=1
+	// +kubebuilder:validation:Optional
+	// +optional
+	Master string `json:"master,omitempty" protobuf:"bytes,2,opt,name=master"`
```

Comment:
optional and minLength=1 is odd. Does empty string mean anything special?

Reply:
what I meant here was that I want master as an optional field, however if you do set it, then you need to provide at least one character in the string.

Reply:
We don't handle empty string and unspecified differently. Would drop minLength.

Reply:
done.